Support fp8 in AOTInductor + support optional<> in C ABI

int3 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113276
* __->__ #112527

This was originally ipiszy's PR: https://github.com/pytorch/pytorch/pull/112358

It turns out that we need to add support for optional types in order to
support fp8 gemm (i.e. scaled_mm). Since our ABI-stable C interface
can't support optional<> directly, I am passing in optional types via
pointer instead.

`AtenTensorHandle`s are already pointers, so nothing needs to change
there. Only value types need to change.

We decided on this approach instead of adding an extra `bool` param to
the callee because this simplifies things. Having the same number of
arguments regardless of whether we are emitting Python / C++ /
ABI-compatible C++ makes codegen easier.

There are a number of existing ABI-compatible functions that have
optional-typed value parameters. Previously, they just assumed they
would never be passed a `nullopt` / `None` at runtime. Changing them to
use pointer types now would break ABI stability, so I have created an
exclude list for those functions.

Finally, I think the current implementation is kind of messy, and only
works for FallbackKernels, even though technically ExternKernels could
also have the same issue. It also doesn't support optional types nested
in lists. I've left FIXME comments for both issues.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

Differential Revision: [D51084289](https://our.internmc.facebook.com/intern/diff/D51084289)

int3 committed:
Support fp8 in AOTInductor + support optional<> in C ABI

This was originally @ipiszy's PR: https://github.com/pytorch/pytorch/pull/112358

It turns out that we need to add support for optional types in order to
support fp8 gemm (i.e. scaled_mm). Since our ABI-stable C interface
can't support optional<> directly, I have created a ShimOptional struct
instead.

This ShimOptional is used only for non-pointer optional types; pointer
optionals can have their nullopt value represented by `nullptr`.

I decided to create a ShimOptional instead of adding an extra `bool`
param to the callee because this simplifies things. Having the same
number of arguments regardless of whether we are emitting Python / C++ /
ABI-compatible C++ makes codegen easier.

There are a number of existing ABI-compatible functions that have
optional-typed parameters. Previously, they just assumed they would
never be passed a `nullopt` / `None` at runtime. Changing them to use
ShimOptional now would break ABI stability, so I have created an
exclude list for those functions.

Finally, I think the current implementation is kind of messy, pulling in
argument type info from a variety of places, and possibly missing some
edge cases with const arg codegen. I've left a bunch of FIXME comments;
would appreciate feedback on whether I could improve things.

[ghstack-poisoned]

int3 commented on torch/_inductor/ir.py:
```diff
@@ -3542,7 +3542,8 @@ def apply_constraint(self):
 
     def codegen_const_args(self):
-        return map(V.graph.wrapper_code.val_to_arg_str, self.constant_args)
+        # FIXME: separate invocation for cpp arg strs?
```
maybe this is fine since looking at its caller, `codegen_args`, it looks like the codegen is meant to target Python only

I'm overall confused by the difference between ExternKernels and FallbackKernels though, and how codegen logic is split between the two. Both of them define `codegen_args`, but they share a single implementation of `codegen_kwargs`...

ipiszy commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -74,6 +74,11 @@ using AOTITorchError = int32_t;
 
+struct ShimOptional {
+  bool has_value;
+  int64_t value;
```
This is not general to other types?

int3 replied:
It's general, I just picked a random 64-bit type. The actual type gets casted here: https://github.com/pytorch/pytorch/pull/112527/files#diff-b60511be1e7fafc2c45e7c0cb3e769ad48b2a1060a69759f58979ffc33b38a79R53

ipiszy commented:
Thanks @int3 !

I wonder have we considered using a pointer to represent optional instead of introducing a boolean value? Code would look cleaner in this case.

wrt existing APIs which are affected by this change: can we add new APIs and deprecate old APIs in parallel? e.g. We could add a new set of APIs in this PR, and then after the PR is released in prod, we remove the legacy branching logics and hard code kernel names of these operators with new API names.

int3 commented:
> I wonder have we considered using a pointer to represent optional instead of introducing a boolean value? Code would look cleaner in this case.

Not sure what you have in mind here. You mean something like `foo(int *opt_int)` that gets invoked like `foo(new int(123))`? Not sure how that's cleaner...

> wrt existing APIs which are affected by this change: can we add new APIs and deprecate old APIs in parallel? e.g. We could add a new set of APIs in this PR, and then after the PR is released in prod, we remove the legacy branching logics and hard code kernel names of these operators with new API names.

Yeah we should do that. I think it might be easier to do it in a follow-up PR though

chenyang78 commented:
> > I wonder have we considered using a pointer to represent optional instead of introducing a boolean value? Code would look cleaner in this case.
> 
> Not sure what you have in mind here. You mean something like `foo(int *opt_int)` that gets invoked like `foo(new int(123))`? Not sure how that's cleaner...
> 

I also think using pointers might be better. The idea would be that we always pass a pointer of the element type of an optional argument, where 

(1) we use NULL to present c10::nullopt; and
(2) make a non-NULL pointer point to any non-nullopt value

For example, for ```c10::optional<int>``` case, we would have something like

```
foo(0 /*c10::nullopt*/);
```

or
```
int arg = 10;
foo(&arg);
```

Because we always know the element type of the optional argument in shim, we wouldn't have ambiguity. Moreover, we wouldn't have to implement our own Optional struct to handle various types.

chenyang78 commented on torch/_inductor/ir.py:
```diff
@@ -4166,14 +4179,21 @@ def is_not_write(arg):
             x.name for x in kernel._schema.arguments if x.kwarg_only
         ]
 
+    def is_legacy_abi_kernel(self):
```
Not sure if we really need this. For these two kernels, all optional arguments have real values, so we wouldn't hit the c10::nullopt path. Moreover, we could add another version of interface, e.g. ```aoti_torch__scaled_dot_product_flash_attention_nullopt``` to handle the missing nullopt cases. Relying on this ```is_legacy_abi_kernel``` looks very hacky to me.

int3 replied:
>  For these two kernels, all optional arguments have real values, so we wouldn't hit the c10::nullopt path.

We would be changing the way the real values get passed in though (regardless of whether we are doing the ShimOptional or pointer approach).

But okay I am happy to do the other less hacky approach once we figure out the issues with getting the valid type above.

ipiszy replied:
Even if we add new versions of these two interfaces (actually we only need to care about the flash_attention one and the other is not used in prod), to make sure the newly published snapshots work with old predictor binaries, we still need to keep the branch until the new API is released in predictor, correct?

desertfire replied:
If `repeat_interleave_Tensor` is not in production, let's fix it by changing its shim API. cc @adnanaziz

int3 replied:
I think you meant to tag @aakhundov :)

Just checked with him, he says it's fine to change

int3 commented on torch/_inductor/ir.py:
```diff
@@ -3581,7 +3585,16 @@ def codegen_kwargs(self):
+                if hasattr(self, "kwargs_default_value"):
+                    type_ = self.kwargs_default_value.get(arg_name).get("type")
+                else:
+                    type_ = None
```
That creates a lot of runtime errors... the issue is that `kwargs_default_value` is defined only on `FallbackKernel`, but this method is defined on `ExternKernel`. I was hoping @desertfire or @jansel might be able to suggest a better way of getting the type here, and/or if this is a thing that I need to concern myself with for non-fallback `ExternKernel`s

int3 commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -182,6 +187,18 @@ AOTI_TORCH_EXPORT AOTITorchError aoti_torch__scaled_dot_product_flash_attention(
+    AtenTensorHandle scale_a,
+    AtenTensorHandle scale_b,
+    AtenTensorHandle scale_result,
+    bool use_fast_accum,
```
Just curious, why not? Is it because C bools and C++ bools are subtly different?

int3 replied:
> I don't think this is sufficient. What if the call passes in a float?

The idea is to use `reinterpret_cast` so the type doesn't matter. Although I see that I forgot to add that into the codegen. But I see what @chenyang78 meant by using pointers instead... that might be nicer, I'll give it a shot.

int3 replied:
Looks like `aoti_torch__scaled_dot_product_flash_attention` is also using `bool`s

int3 replied:
https://stackoverflow.com/questions/40020423/getting-bool-from-c-to-c-and-back sounds like this is relatively safe

desertfire replied:
As described in that stackoverflow thread, "C's and C++'s bool type are different, but, as long as you stick to the same compiler (in your case, gcc), it should be safe, as this is a reasonable common scenario.", but we can't make that assumption here, as we don't know what the `model.so` was compiled with.

int3 replied:
Oh I thought we were only concerned with models compiled under fbcode. Okay, I'll change it.

int3 committed:
Update on "Support fp8 in AOTInductor + support optional<> in C ABI"


This was originally ipiszy's PR: https://github.com/pytorch/pytorch/pull/112358

It turns out that we need to add support for optional types in order to
support fp8 gemm (i.e. scaled_mm). Since our ABI-stable C interface
can't support optional<> directly, I have created a ShimOptional struct
instead.

This ShimOptional is used only for non-pointer optional types; pointer
optionals can have their nullopt value represented by `nullptr`.

I decided to create a ShimOptional instead of adding an extra `bool`
param to the callee because this simplifies things. Having the same
number of arguments regardless of whether we are emitting Python / C++ /
ABI-compatible C++ makes codegen easier.

There are a number of existing ABI-compatible functions that have
optional-typed parameters. Previously, they just assumed they would
never be passed a `nullopt` / `None` at runtime. Changing them to use
ShimOptional now would break ABI stability, so I have created an
exclude list for those functions.

Finally, I think the current implementation is kind of messy, pulling in
argument type info from a variety of places, and possibly missing some
edge cases with const arg codegen. I've left a bunch of FIXME comments;
would appreciate feedback on whether I could improve things.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

int3 committed:
Update on "Support fp8 in AOTInductor + support optional<> in C ABI"


This was originally ipiszy's PR: https://github.com/pytorch/pytorch/pull/112358

It turns out that we need to add support for optional types in order to
support fp8 gemm (i.e. scaled_mm). Since our ABI-stable C interface
can't support optional<> directly, I am passing in optional types via
pointer instead.

`AtenTensorHandle`s are already pointers, so nothing needs to change
there. Only value types need to change.

We decided on this approach instead of adding an extra `bool` param to
the callee because this simplifies things. Having the same number of
arguments regardless of whether we are emitting Python / C++ /
ABI-compatible C++ makes codegen easier.

There are a number of existing ABI-compatible functions that have
optional-typed value parameters. Previously, they just assumed they
would never be passed a `nullopt` / `None` at runtime. Changing them to
use pointer types now would break ABI stability, so I have created an
exclude list for those functions.

Finally, I think the current implementation is kind of messy, and only
works for FallbackKernels, even though technically ExternKernels could
also have the same issue. It also doesn't support optional types nested
in lists. I've left FIXME comments for both issues.

[ghstack-poisoned]

int3 commented:
* Now using @chenyang78's suggestion re passing all optionals by pointer
* Not trying to handle optional types in ExternKernels for now; only FallbackKernels. I think we can make ExternKernels aware of cpp schemas too and further simplify the code, but that is future work.
* Not doing the legacy ABI versioning in this diff; @chenyang78 says he will tackle it as he ran into something else that needs this too
* Still using `bool`s in the interface for now, pending feedback. Initial research suggests that this is safe

ipiszy commented on torch/_inductor/lowering.py:
```diff
@@ -2018,7 +2018,6 @@ def apply_constraint(arg, fx_arg):
 make_fallback(aten._thnn_fused_lstm_cell, require_dense)
 make_fallback(aten.topk)
 make_fallback(aten.upsample_bicubic2d_backward, require_contiguous)
-make_fallback(aten._scaled_mm.default)
```
I feel we probably need to keep this since we haven't added lowering for _scaled_mm? (I'm a bit confused about this as well). Have you tried that `test/inductor/test_fp8.py` can run successfully if we remove this fallback?

int3 replied:
yep, I've checked and the test passes

not entirely sure either about how the fallback mechanism works though

int3 committed:
Update on "Support fp8 in AOTInductor + support optional<> in C ABI"


This was originally ipiszy's PR: https://github.com/pytorch/pytorch/pull/112358

It turns out that we need to add support for optional types in order to
support fp8 gemm (i.e. scaled_mm). Since our ABI-stable C interface
can't support optional<> directly, I am passing in optional types via
pointer instead.

`AtenTensorHandle`s are already pointers, so nothing needs to change
there. Only value types need to change.

We decided on this approach instead of adding an extra `bool` param to
the callee because this simplifies things. Having the same number of
arguments regardless of whether we are emitting Python / C++ /
ABI-compatible C++ makes codegen easier.

There are a number of existing ABI-compatible functions that have
optional-typed value parameters. Previously, they just assumed they
would never be passed a `nullopt` / `None` at runtime. Changing them to
use pointer types now would break ABI stability, so I have created an
exclude list for those functions.

Finally, I think the current implementation is kind of messy, and only
works for FallbackKernels, even though technically ExternKernels could
also have the same issue. It also doesn't support optional types nested
in lists. I've left FIXME comments for both issues.

[ghstack-poisoned]

chenyang78 commented on torch/_inductor/ir.py:
```diff
@@ -4185,21 +4204,30 @@ def __repr__(self):
+                )
+                for param, x in zip(self.op_overload._schema.arguments, args)
+            ]
+            if hasattr(self, "args_default_value") and not config.is_fbcode():
```
Seems this PR is not rebased to the head of the master? This part is different from:

https://github.com/pytorch/pytorch/blob/main/torch/_inductor/ir.py#L4265-L4285

where ```config.is_fbcode()``` has been removed.

chenyang78 commented on torch/_meta_registrations.py:
```diff
@@ -1891,6 +1891,31 @@ def meta_mm(a, b):
     return a.new_empty(N, P)
 
 
+@register_meta([aten._scaled_mm.default])
```
Looks like a similar change has been merged: https://github.com/pytorch/pytorch/pull/112609

int3 committed:
Update on "Support fp8 in AOTInductor + support optional<> in C ABI"


This was originally ipiszy's PR: https://github.com/pytorch/pytorch/pull/112358

It turns out that we need to add support for optional types in order to
support fp8 gemm (i.e. scaled_mm). Since our ABI-stable C interface
can't support optional<> directly, I am passing in optional types via
pointer instead.

`AtenTensorHandle`s are already pointers, so nothing needs to change
there. Only value types need to change.

We decided on this approach instead of adding an extra `bool` param to
the callee because this simplifies things. Having the same number of
arguments regardless of whether we are emitting Python / C++ /
ABI-compatible C++ makes codegen easier.

There are a number of existing ABI-compatible functions that have
optional-typed value parameters. Previously, they just assumed they
would never be passed a `nullopt` / `None` at runtime. Changing them to
use pointer types now would break ABI stability, so I have created an
exclude list for those functions.

Finally, I think the current implementation is kind of messy, and only
works for FallbackKernels, even though technically ExternKernels could
also have the same issue. It also doesn't support optional types nested
in lists. I've left FIXME comments for both issues.

[ghstack-poisoned]

int3 commented:
* Rebased
* Ensured that we can correctly codegen references to `torch.float8_*` types in ABI-compatible mode

int3 committed:
Update on "Support fp8 in AOTInductor + support optional<> in C ABI"


This was originally ipiszy's PR: https://github.com/pytorch/pytorch/pull/112358

It turns out that we need to add support for optional types in order to
support fp8 gemm (i.e. scaled_mm). Since our ABI-stable C interface
can't support optional<> directly, I am passing in optional types via
pointer instead.

`AtenTensorHandle`s are already pointers, so nothing needs to change
there. Only value types need to change.

We decided on this approach instead of adding an extra `bool` param to
the callee because this simplifies things. Having the same number of
arguments regardless of whether we are emitting Python / C++ /
ABI-compatible C++ makes codegen easier.

There are a number of existing ABI-compatible functions that have
optional-typed value parameters. Previously, they just assumed they
would never be passed a `nullopt` / `None` at runtime. Changing them to
use pointer types now would break ABI stability, so I have created an
exclude list for those functions.

Finally, I think the current implementation is kind of messy, and only
works for FallbackKernels, even though technically ExternKernels could
also have the same issue. It also doesn't support optional types nested
in lists. I've left FIXME comments for both issues.

[ghstack-poisoned]

int3 committed:
Update on "Support fp8 in AOTInductor + support optional<> in C ABI"


This was originally ipiszy's PR: https://github.com/pytorch/pytorch/pull/112358

It turns out that we need to add support for optional types in order to
support fp8 gemm (i.e. scaled_mm). Since our ABI-stable C interface
can't support optional<> directly, I am passing in optional types via
pointer instead.

`AtenTensorHandle`s are already pointers, so nothing needs to change
there. Only value types need to change.

We decided on this approach instead of adding an extra `bool` param to
the callee because this simplifies things. Having the same number of
arguments regardless of whether we are emitting Python / C++ /
ABI-compatible C++ makes codegen easier.

There are a number of existing ABI-compatible functions that have
optional-typed value parameters. Previously, they just assumed they
would never be passed a `nullopt` / `None` at runtime. Changing them to
use pointer types now would break ABI stability, so I have created an
exclude list for those functions.

Finally, I think the current implementation is kind of messy, and only
works for FallbackKernels, even though technically ExternKernels could
also have the same issue. It also doesn't support optional types nested
in lists. I've left FIXME comments for both issues.

[ghstack-poisoned]

chenyang78 reviewed (approved):
LGTM. Thanks!

int3 committed:
Update on "Support fp8 in AOTInductor + support optional<> in C ABI"


This was originally ipiszy's PR: https://github.com/pytorch/pytorch/pull/112358

It turns out that we need to add support for optional types in order to
support fp8 gemm (i.e. scaled_mm). Since our ABI-stable C interface
can't support optional<> directly, I am passing in optional types via
pointer instead.

`AtenTensorHandle`s are already pointers, so nothing needs to change
there. Only value types need to change.

We decided on this approach instead of adding an extra `bool` param to
the callee because this simplifies things. Having the same number of
arguments regardless of whether we are emitting Python / C++ /
ABI-compatible C++ makes codegen easier.

There are a number of existing ABI-compatible functions that have
optional-typed value parameters. Previously, they just assumed they
would never be passed a `nullopt` / `None` at runtime. Changing them to
use pointer types now would break ABI stability, so I have created an
exclude list for those functions.

Finally, I think the current implementation is kind of messy, and only
works for FallbackKernels, even though technically ExternKernels could
also have the same issue. It also doesn't support optional types nested
in lists. I've left FIXME comments for both issues.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

desertfire commented on torch/_inductor/lowering.py:
```diff
@@ -2087,7 +2087,6 @@ def is_aligned(x):
 make_fallback(aten._thnn_fused_lstm_cell, require_dense)
 make_fallback(aten.topk)
 make_fallback(aten.upsample_bicubic2d_backward, require_contiguous)
-make_fallback(aten._scaled_mm.default, constrain_to_fx_strides)
```
You are not lowering `_scaled_mm` in this PR, right? Removing `_scaled_mm` here will just make it use a silent implicit fallback.

int3 replied:
ah yeah I wasn't sure about this (@ipiszy was asking about it too above). I'll put it back

chenyang78 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -2281,10 +2281,28 @@ def extract_output_name(out):
+            and isinstance(type_, torch.OptionalType)
+        ):
+            if val is None:
+                return "nullptr"
```
Perhaps we should just use ```0``` instead of ```nullptr```, which is a C++ construct.

int3 committed:
Update on "Support fp8 in AOTInductor + support optional<> in C ABI"


This was originally ipiszy's PR: https://github.com/pytorch/pytorch/pull/112358

It turns out that we need to add support for optional types in order to
support fp8 gemm (i.e. scaled_mm). Since our ABI-stable C interface
can't support optional<> directly, I am passing in optional types via
pointer instead.

`AtenTensorHandle`s are already pointers, so nothing needs to change
there. Only value types need to change.

We decided on this approach instead of adding an extra `bool` param to
the callee because this simplifies things. Having the same number of
arguments regardless of whether we are emitting Python / C++ /
ABI-compatible C++ makes codegen easier.

There are a number of existing ABI-compatible functions that have
optional-typed value parameters. Previously, they just assumed they
would never be passed a `nullopt` / `None` at runtime. Changing them to
use pointer types now would break ABI stability, so I have created an
exclude list for those functions.

Finally, I think the current implementation is kind of messy, and only
works for FallbackKernels, even though technically ExternKernels could
also have the same issue. It also doesn't support optional types nested
in lists. I've left FIXME comments for both issues.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

int3 commented:
* put back explicit fallback
* remove `repeat_interleave_Tensor` from legacy kernel list
* the above revealed that `pointer_to_optional` wasn't handling implicit conversions, so I fixed that too

int3 committed:
Update on "Support fp8 in AOTInductor + support optional<> in C ABI"


This was originally ipiszy's PR: https://github.com/pytorch/pytorch/pull/112358

It turns out that we need to add support for optional types in order to
support fp8 gemm (i.e. scaled_mm). Since our ABI-stable C interface
can't support optional<> directly, I am passing in optional types via
pointer instead.

`AtenTensorHandle`s are already pointers, so nothing needs to change
there. Only value types need to change.

We decided on this approach instead of adding an extra `bool` param to
the callee because this simplifies things. Having the same number of
arguments regardless of whether we are emitting Python / C++ /
ABI-compatible C++ makes codegen easier.

There are a number of existing ABI-compatible functions that have
optional-typed value parameters. Previously, they just assumed they
would never be passed a `nullopt` / `None` at runtime. Changing them to
use pointer types now would break ABI stability, so I have created an
exclude list for those functions.

Finally, I think the current implementation is kind of messy, and only
works for FallbackKernels, even though technically ExternKernels could
also have the same issue. It also doesn't support optional types nested
in lists. I've left FIXME comments for both issues.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel commented:
I will let @ipiszy review this.  Let me know if there is something here for me specifically to look at.

int3 commented:
@ghimport

ipiszy commented:
@ipiszy has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D51084289).

int3 committed:
Update on "Support fp8 in AOTInductor + support optional<> in C ABI"


This was originally ipiszy's PR: https://github.com/pytorch/pytorch/pull/112358

It turns out that we need to add support for optional types in order to
support fp8 gemm (i.e. scaled_mm). Since our ABI-stable C interface
can't support optional<> directly, I am passing in optional types via
pointer instead.

`AtenTensorHandle`s are already pointers, so nothing needs to change
there. Only value types need to change.

We decided on this approach instead of adding an extra `bool` param to
the callee because this simplifies things. Having the same number of
arguments regardless of whether we are emitting Python / C++ /
ABI-compatible C++ makes codegen easier.

There are a number of existing ABI-compatible functions that have
optional-typed value parameters. Previously, they just assumed they
would never be passed a `nullopt` / `None` at runtime. Changing them to
use pointer types now would break ABI stability, so I have created an
exclude list for those functions.

Finally, I think the current implementation is kind of messy, and only
works for FallbackKernels, even though technically ExternKernels could
also have the same issue. It also doesn't support optional types nested
in lists. I've left FIXME comments for both issues.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D51084289](https://our.internmc.facebook.com/intern/diff/D51084289)

[ghstack-poisoned]

int3 committed:
Update on "Support fp8 in AOTInductor + support optional<> in C ABI"


This was originally ipiszy's PR: https://github.com/pytorch/pytorch/pull/112358

It turns out that we need to add support for optional types in order to
support fp8 gemm (i.e. scaled_mm). Since our ABI-stable C interface
can't support optional<> directly, I am passing in optional types via
pointer instead.

`AtenTensorHandle`s are already pointers, so nothing needs to change
there. Only value types need to change.

We decided on this approach instead of adding an extra `bool` param to
the callee because this simplifies things. Having the same number of
arguments regardless of whether we are emitting Python / C++ /
ABI-compatible C++ makes codegen easier.

There are a number of existing ABI-compatible functions that have
optional-typed value parameters. Previously, they just assumed they
would never be passed a `nullopt` / `None` at runtime. Changing them to
use pointer types now would break ABI stability, so I have created an
exclude list for those functions.

Finally, I think the current implementation is kind of messy, and only
works for FallbackKernels, even though technically ExternKernels could
also have the same issue. It also doesn't support optional types nested
in lists. I've left FIXME comments for both issues.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D51084289](https://our.internmc.facebook.com/intern/diff/D51084289)

[ghstack-poisoned]

int3 committed:
Update on "Support fp8 in AOTInductor + support optional<> in C ABI"


This was originally ipiszy's PR: https://github.com/pytorch/pytorch/pull/112358

It turns out that we need to add support for optional types in order to
support fp8 gemm (i.e. scaled_mm). Since our ABI-stable C interface
can't support optional<> directly, I am passing in optional types via
pointer instead.

`AtenTensorHandle`s are already pointers, so nothing needs to change
there. Only value types need to change.

We decided on this approach instead of adding an extra `bool` param to
the callee because this simplifies things. Having the same number of
arguments regardless of whether we are emitting Python / C++ /
ABI-compatible C++ makes codegen easier.

There are a number of existing ABI-compatible functions that have
optional-typed value parameters. Previously, they just assumed they
would never be passed a `nullopt` / `None` at runtime. Changing them to
use pointer types now would break ABI stability, so I have created an
exclude list for those functions.

Finally, I think the current implementation is kind of messy, and only
works for FallbackKernels, even though technically ExternKernels could
also have the same issue. It also doesn't support optional types nested
in lists. I've left FIXME comments for both issues.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

Differential Revision: [D51084289](https://our.internmc.facebook.com/intern/diff/D51084289)

[ghstack-poisoned]

int3 commented:
@pytorchbot merge

int3 commented:
@pytorchbot merge

int3 commented:
Let's see if unlinking works...


@pytorchbot merge

swolchok commented:
> Changing them to use pointer types now would break ABI stability

are we stuck with ABI stability already?

int3 commented:
We do need to maintain backwards compatibility for `_scaled_dot_product_flash_attention` which is being used. But I believe the plan is to make a temporary `v2` version of the function, deprecate the old one, then move all callers to using the new ABI format. @chenyang78 is looking into it.

chenyang78 commented:
> We do need to maintain backwards compatibility for `_scaled_dot_product_flash_attention` which is being used. But I believe the plan is to make a temporary `v2` version of the function, deprecate the old one, then move all callers to using the new ABI format. @chenyang78 is looking into it.

@swolchok https://github.com/pytorch/pytorch/pull/113487
